packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m

```diff
@@ -0,0 +1,285 @@
+{
+  [self _setRCTErrorHandlers];
+}
+
+- (void)handleError:(NSError *)error
+{
+  [self _startPipelineWithEncounteredError:error];
+}
+
+- (void)handleException:(NSException *)exception
+{
+  [self _startPipelineWithEncounteredError:exception];
+}
+
+- (void)notifyNewRemoteLoadStatus:(EXUpdatesRemoteLoadStatus)newStatus
+{
+  dispatch_async(_errorRecoveryQueue, ^{
+    if (!self->_isWaitingForRemoteUpdate) {
+      return;
+    }
+    self->_isWaitingForRemoteUpdate = NO;
+    if (newStatus == EXUpdatesRemoteLoadStatusNewUpdateLoaded) {
+      [self->_pipeline insertObject:@(EXUpdatesErrorRecoveryTaskLaunchNew) atIndex:0];
+    }
+    [self _runNextTask];
+  });
+}
+
+# pragma mark - pipeline tasks
+
+- (void)_startPipelineWithEncounteredError:(id)encounteredError
+{
+  dispatch_async(_errorRecoveryQueue, ^{
+    [self->_encounteredErrors addObject:encounteredError];
+
+    if (self->_delegate.launchedUpdate.successfulLaunchCount > 0) {
+      [self->_pipeline removeObject:@(EXUpdatesErrorRecoveryTaskLaunchCached)];
+    } else if (!self->_rctContentHasAppeared) {
+      [self->_delegate markFailedLaunchForLaunchedUpdate];
+    }
+
+    if (!self->_isRunning) {
+      self->_isRunning = YES;
+      [self _runNextTask];
+    }
+  });
+}
+
+- (void)_runNextTask
+{
+  dispatch_assert_queue(_errorRecoveryQueue);
+  NSNumber *nextTask = _pipeline[0];
+  [_pipeline removeObjectAtIndex:0];
+  switch ((EXUpdatesErrorRecoveryTask)nextTask.integerValue) {
+    case EXUpdatesErrorRecoveryTaskWaitForRemoteUpdate:
+      [self _waitForRemoteLoaderToFinish];
+      break;
+        // EXUpdatesErrorRecoveryTaskLaunchNew is called only after a new update is downloaded
+        // and added to the cache, so it equivalent to EXUpdatesErrorRecoveryTaskLaunchCached
+    case EXUpdatesErrorRecoveryTaskLaunchNew:
+    case EXUpdatesErrorRecoveryTaskLaunchCached:
+      [self _tryRelaunchFromCache];
+      break;
+    case EXUpdatesErrorRecoveryTaskCrash:
+      [self _crash];
+      break;
+    default:
+      @throw [NSException exceptionWithName:NSInternalInconsistencyException reason:@"EXUpdatesErrorRecovery cannot run the next task" userInfo:@{@"nextTaskValue": nextTask}];
+  }
+}
+
+- (void)_waitForRemoteLoaderToFinish
+{
+  dispatch_assert_queue(_errorRecoveryQueue);
+  if (_delegate.remoteLoadStatus == EXUpdatesRemoteLoadStatusLoading) {
+    _isWaitingForRemoteUpdate = YES;
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, _remoteLoadTimeout * NSEC_PER_MSEC), _errorRecoveryQueue, ^{
+      if (!self->_isWaitingForRemoteUpdate) {
+        return;
+      }
+      self->_isWaitingForRemoteUpdate = NO;
+      [self _runNextTask];
+    });
+    return;
+  } else if (_delegate.remoteLoadStatus == EXUpdatesRemoteLoadStatusNewUpdateLoaded) {
+    [_pipeline insertObject:@(EXUpdatesErrorRecoveryTaskLaunchNew) atIndex:0];
+    [self _runNextTask];
+  } else {
+    _isWaitingForRemoteUpdate = YES;
+    [_delegate loadRemoteUpdate];
+    [self _runNextTask];
```

Comment:
While the delegate is loading the update, won't this run the EXUpdatesErrorRecoveryTaskLaunchCached task right away? Two things: this case looks like it isn't waiting for the remote loader to finish (in a method named `_waitForRemoteLoaderToFinish`), and shouldn't this look for `checkAutomatically: ON_ERROR_RECOVERY` or similar instead of always looking for an update?

Reply:
Re: running the next task, nice catch, that indeed shouldn't be here.

Re: `checkAutomatically: ON_ERROR_RECOVERY` -- this code won't run at all if updates are disabled entirely, but in the managed workflow we only allow two settings for `checkAutomatically`, `ALWAYS` and `ON_ERROR_RECOVERY`, and in both cases we would check for an update here. Do you think there's value in adding a third setting where updates are enabled but no automatic check happens, even in an error recovery situation?